syntax = "proto3";

package analysis.dvvp.proto;


enum EumStatus {
    SUCCESS = 0;
    FAILED = 1;
}

message Header {
    string Magic = 1;
    string job_ctx = 2;
}

// for transfer data and flush to file
message FileChunkReq {
    Header Hdr = 1;
    string FileName = 2;
    int64 Offset = 3;
    bytes Chunk = 4;
    int32 ChunkSizeInBytes = 5;
    bool IsLastChunk = 6;
    bool NeedAck = 7;
    string Tag = 8;
    string TagSuffix = 9;
    uint64 ChunkStartTime = 10;
    uint64 ChunkEndTime = 11;
    int32 DataModule = 12;
}

message FileChunkFlushReq {
    Header Hdr = 1;
    string FileName = 2;
    int32 DataModule = 3;
}

message FileAck {
    Header Hdr = 1;
    string FileName = 2;
    EumStatus Status = 3;
    int32 SizeWritten = 4;
    string Message = 5;
}

//common response
message Response {
    Header Hdr = 1;
    string JobId = 2;
    EumStatus Status = 3;
    string Message = 4;
}

// for remote job start
message RemoteJobStartReq {
    int32 devId = 1;
    string SampleConfig = 2;
}

// for job start/stop
message JobStartReq {
    Header Hdr = 1;
    string SampleConfig = 2;
}

message JobStopReq {
    Header Hdr = 1;
    string JobId = 2;
}

message ReplayStartReq {
    Header Hdr = 1;
    int32 ReplayId = 2;
    repeated string ctrl_cpu_events = 3;
    repeated string ts_cpu_events = 4;
    repeated string ai_cpu_events = 5;
    repeated string ai_core_events = 6;
    repeated int32 ai_core_events_cores = 7;    //ai_core_events_cores used only when sample based
    repeated string llc_events = 8;
    repeated string ddr_events = 9;
    repeated string aiv_events = 10;
    repeated int32 aiv_events_cores = 11;   //ai_vecotor_core_events_cores used only when sample based
}

message ReplayStopReq {
    Header Hdr = 1;
    int32 ReplayId = 2;
}

// AI Host will send FinishJob to UI host when sampling finished
message FinishJobRsp {
    Header Hdr = 1;
    string JobId = 2;
    EumStatus Status = 3;
    string Message = 4;
}

// handshake between host and device
message CtrlChannelHandshake {
    string jobId = 1;
    int32 devId = 2;
    string mode = 3; // system-wide or def_mode or ""
}

message DataChannelHandshake {
    string jobId = 1;
    int32 devId = 2;
    string mode = 3;
}

// sync device list
message DeviceInfo {
    string dev_id = 1;
}

message SyncDeviceList {
    Header Hdr = 1;
    repeated DeviceInfo dev_list = 2;
}

message DataChannelFinish {
    Header Hdr = 1;
}

message TypeEvents {
    string type = 1;
    string events = 2;
};

message ProfilerTypeEvents {
    repeated TypeEvents aicoreMetrics = 1;
};

message ProfilerEvents {
    string ai_core_events = 1;
    string L2_cache_events = 2;
};

message JobCtx {
    bool stream_enabled = 1;
    string result_dir = 2;
    string format = 3; //json, binary
}
message ModuleConfig {
    map<string, string> switches = 1;
}
message ProfilerJobConfig {
    JobCtx job_ctx = 1;
    map<string, ModuleConfig> modules = 2;
}

message ProfilerConf {
    string aicoreMetrics = 1;
    uint32 aicoreSamplingInterval = 2;
    uint32 cpuSamplingInterval = 3;
    uint32 sysSamplingInterval = 4;
    uint32 appSamplingInterval = 5;
    uint32 hardwareMemSamplingInterval = 6;
    uint32 ioSamplingInterval = 7;
    uint32 interconnectionSamplingInterval = 8;
    uint32 dvppSamplingInterval = 9;
    repeated ProfilerEvents events = 10;
    string aivMetrics = 11;
    uint32 aivSamplingInterval = 12;
}

message ProfilerFeature {
    string name = 1;
    repeated ProfilerConf conf = 2;
}

message ProfilerFeatureCtx {
    string jobID = 1;
    string traceId = 2;
    string resultPath = 3;
    repeated string deviceID = 4;   // ["all"] or ["0", "1"]
    repeated ProfilerFeature features= 5;
}

message ProfilerFeatureConfig {
    repeated ProfilerFeatureCtx startCfg= 1;
}

message MsProfStartReq {
    string job_id = 1;      // trace id produced by libmsprof
    string ts_fw_training = 2;
    string rts_profiling = 3;
    string task_base = 4;
    string hwts_log = 5;
    string ts_timeline = 7;
    string ts_task_track = 8;
    string docker_id = 9;
    string job_info = 10;   // job id send by GE
    string feature_name = 11;
    string ai_core_events = 12;
    string l2_cache_events = 13;
    string traceId = 14;
    string result_path = 15;
    string op_trace_conf = 16;
    string system_trace_conf = 17;
    string task_trace_conf = 18;

}
message DataChannelClose {
    Header Hdr = 1;
}

// for generating info.json _________BEGIN
message InfoCpu {
    uint32 Id = 1;
    string Name = 2;
    uint32 Frequency = 3;
    uint32 Logical_CPU_Count = 4;
    string Type = 5;
}

message InfoDeviceInfo {
    uint32 id = 1;
    uint32 env_type = 2;
    string ctrl_cpu_id = 3;
    uint32 ctrl_cpu_core_num = 4;
    uint32 ctrl_cpu_endian_little = 5;
    uint32 ts_cpu_core_num = 6;
    uint32 ai_cpu_core_num = 7;
    uint32 ai_core_num = 8;
    uint32 ai_cpu_core_id = 9;
    uint32 ai_core_id = 10;
    uint32 aicpu_occupy_bitmap = 11;
    string ctrl_cpu = 12;
    string ai_cpu = 13;
    uint32 aiv_num = 14;
    string hwts_frequency = 15;
    string aic_frequency = 16;
    string aiv_frequency = 17;
}

message NetCardInfo {
    string netCardName = 1;
    int32 speed = 2;
}

message InfoMain {
    string jobInfo = 1;
    string mac = 3;
    string OS = 4;
    uint32 cpuCores = 5;
    string hostname = 6;
    string hwtype = 7;
    string devices = 8;
    string platform = 9;
    repeated InfoCpu CPU = 10;
    repeated InfoDeviceInfo DeviceInfo = 11;
    string platform_version = 12;
    string pid = 13;
    uint32 memoryTotal = 14;
    uint32 cpuNums = 15;
    uint32 sysClockFreq = 16;
    string upTime = 17;
    uint32 netCardNums = 18;
    repeated NetCardInfo netCard = 19;
}
// for generating info.json _________END

message AppLog {
    string log = 1;
    string jobId = 2;
}

message CollectionStartEndTime {
    string collectionDateBegin = 1;
    string collectionDateEnd = 2;
    string collectionTimeBegin = 3;
    string collectionTimeEnd = 4;
    string clockMonotonicRaw = 5;
}